packages/react-router-dom/modules/Link.js

```diff
@@ -47,14 +48,12 @@ class Link extends React.Component {
               : to;
           const href = location ? context.history.createHref(location) : "";
 
-          return (
-            <a
-              {...rest}
-              onClick={event => this.handleClick(event, context.history)}
-              href={href}
-              ref={innerRef}
-            />
-          );
+          return React.createElement(component, {
+            ...rest,
+            href,
+            onClick: event => this.handleClick(event, context.history),
```

Comment:
One thing I'm unsure of is if we should pass the `onClick` prop through to the component like this, or something else. For `<a>` this makes sense, but for other elements, they may want a more generic interface that doesn't require them to pass an `event` argument.

Reply:
Yeah this has similar problems in that it controls rendering of the component and assumes the component has the same props API as `a`.

Render props seems like the declarative way to invert control and allow the router's calculated props to be correctly wired up to a composite component. Same for `NavLink`, which makes assumptions about how to style elements rather than exposing the state you need to do that yourself.

Another alternative might be to expose some of the building blocks used here so that people can make their own declarative interface over the router context, rather than using these convenience components.